api/repositories/shared.go

```diff
@@ -59,3 +67,73 @@ func getLabelOrAnnotation(mapObj map[string]string, key string) string {
 	}
 	return mapObj[key]
 }
+
+func GetNamespacedResourceNamespace(ctx context.Context, privilegedClient client.Client, userClientFactory UserK8sClientFactory, authInfo authorization.Info, resourceGUID string, resourceList client.ObjectList) (string, error) {
+	err := privilegedClient.List(ctx, resourceList, client.MatchingFields{"metadata.name": resourceGUID})
+	if err != nil { // untested
+		return "", fmt.Errorf("privileged list failed: %w", err)
+	}
+
+	switch t := resourceList.(type) {
```

Comment:
This huge switch does not look great - it has quite a lot of duplicated code only because there is no `Items` method in the `ObjectList` interface. Maybe we could use reflection to see whether the incoming `resourceList` has an `Items` field and return error otherwise. Or, we could use some code generator as you suggested, but that might be way too complex.

Reply:
The reason for the switch is to avoid reflection. I treat reflection as a last resort - hence this approach to some unfortunate but maintainable code that should be relatively set and forget.

Reply:
Further digging turns up the [kubernetes apimachinery converter](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/converter.go). This looks like it'll help us out in this situation to get to unstructured objects. It uses reflection under the hood, so maybe this is the best way to do it. Since kubernetes provides this for us, I'd recommend we look into using the provided tooling instead of the reflect library directly.

Reply:
Further spelunking suggests that https://pkg.go.dev/k8s.io/client-go/dynamic might help us.

Reply:
Yes, `dynamic` seems to do the trick! We'd need a privileged [`dynamic.Interface`](https://pkg.go.dev/k8s.io/client-go/dynamic#Interface) client (not sure if in substitution of `privilegedClient` or in addition to it), and then we could get an [`UnstructuredList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#UnstructuredList) which _should_ be enough to check the length and to extract the namespace. [`Unstructured`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured) objects are essentially `map[string]interface{}`s but it should work!